tests/test_forms.py

```diff
@@ -0,0 +1,190 @@
+            dict_to_struct({"id": "field0", "type": "Integer", "value": 2}),
+            dict_to_struct({"id": "field1", "type": "Float", "value": 2.5}),
+        ]
+        process_form_fields(fields, {"template_in": "eggs",
+                                     "template_out": "choice1",
+                                     "field0": "1",
+                                     "field1": "1.5",
+                                     })
+        _ = CreateForm(fields)
+        self.assertEqual(fields[0].value, "eggs")
+        self.assertEqual(fields[1].value, "choice1")
+        self.assertEqual(fields[2].value, 1)
+        self.assertEqual(fields[3].value, 1.5)
+
+    def test_invalid_data(self):
+        fields = [
+            dict_to_struct({"id": "template_in", "type": "Text", "value": "spam"}),
+            dict_to_struct({"id": "template_out", "type": "Text", "value": "spam"}),
+            dict_to_struct({"id": "field0", "type": "Integer", "value": 2}),
+        ]
+
+        expected_error_msg = (
+            "form field 'field0' value 'a' is not a 'Integer'.")
+        with self.assertRaises(ValidationError) as cm:
+            process_form_fields(fields, {"template_in": "eggs",
+                                         "template_out": "choice1",
+                                         "field0": "a",
+                                         })
+        self.assertIn(expected_error_msg, str(cm.exception))
+
+    def test_create_form(self):
+        fields = [
+            dict_to_struct({"id": "template_in", "type": "Text", "value": "spam"}),
+            dict_to_struct({"id": "template_out", "type": "Text",
+                            "value": "out.yml"}),
+            dict_to_struct({"id": "field0", "type": "Integer", "value": 2}),
+            dict_to_struct({"id": "field1", "type": "Float", "value": 2.5}),
+            dict_to_struct({"id": "field2", "type": "Choice",
+                            "choices": ["choice1", "~DEFAULT~ choice2"]}),
+            dict_to_struct({"id": "field3", "type": "Hidden", "value": 2}),
+        ]
+        process_form_fields(fields, {})
+        form = CreateForm(fields)
+        for field, ftype in [("field0", forms.IntegerField),
+                             ("field1", forms.FloatField),
+                             ("field2", forms.ChoiceField),
+                             ("template_in", forms.CharField),
+                             ("template_out", forms.CharField)]:
+            self.assertIn(field, form.fields)
+            self.assertIsInstance(form.fields[field], ftype)
+
+        self.assertNotIn("field3", form.fields)
+        # Check that template_out has id appended
+        self.assertEqual(form.template_out, "out_{}.yml".format(form.id))
+        self.assertIn(form.id, form.get_jinja_text()[0])
+
+
+class FormsBase(SingleCourseTestMixin, MockAddMessageMixing, TestCase):
+
+    initial_commit_sha = "f3e9d31a61714e759a6ea12b900b173accb753f5"
```

Comment:
@inducer, I used this commit from https://github.com/inducer/relate-sample/pull/12 which has `flows/instant_flow.jinja` file and it is included by the created flow using `{% include 'flows/instant_flow.jinja' %}`, but when I try this in tests the templating engine can't find this file. How can I make jinja2 see this file?